### ISSUE TITLE
How to use my own dataset for training?

### ISSUE THREAD

**Issue by xiannsa** [cvg/SOLD2#82]
Excuse me, is it true that the method for training my own dataset involves preparing a folder to store images for training,  write a new dataloader by imitating wireframe adding it to dataset_util.py, and then following the provided five steps for training?

**Comment by rpautrat** [comment]
Yes, that is a good summary for re-training on a custom dataset!

**Comment by xiannsa** [comment]
Excuse me, I encountered a problem while running the second step, bash: checkpoint-epoch019-end.tar: There is no such file or directory, but the path and file name I wrote in the command line are correct. Could you give me some advice?

**Comment by xiannsa** [comment]
Excuse me, I imitated Holicity's dataloader, and after the second step of training, no files were generated. So how should I proceed with the third step? I hope you can give some advice.

**Comment by abhiagwl4262** [comment]
@rpautrat 

I converted my dataset in lcnn format to wireframe format using - 
https://github.com/Delay-Xili/F-Clip/blob/master/dataset/wireframe.py &
https://github.com/Delay-Xili/F-Clip/blob/master/dataset/wireframe_line.py

I hope that also works.

**Comment by rpautrat** [comment]
> Excuse me, I imitated Holicity's dataloader, and after the second step of training, no files were generated. So how should I proceed with the third step? I hope you can give some advice.

Did any error happen? Step 2 should be generating the ground truth files. Did you also check that the paths are correct, for example where was the output path that you selected?

**Comment by rpautrat** [comment]
> @rpautrat
> 
> I converted my dataset which is in lcnn format currently to wireframe format using - https://github.com/Delay-Xili/F-Clip/blob/master/dataset/wireframe.py & https://github.com/Delay-Xili/F-Clip/blob/master/dataset/wireframe_line.py
> 
> I hope that also works.

I am not sure to understand the difference between these two formats, but if you use the one of the original Wireframe dataset, this should be fine.

**Comment by xiannsa** [comment]
Thank you for your help,I have completed the third step. However, I have encountered a new issue while training in the fourth step: "Input type (torch.cuda.DoubleTensor) and weight type (torch.cuda.FloatTensor) should be the same.

**Comment by rpautrat** [comment]
This means that there is a mismatch of types: the model weights are torch.float, while some input (probably your input images) are torch.double. You need to convert your images to float with `images.float()`.

**Comment by xiannsa** [comment]
Hello, thank you for your help earlier. I was able to successfully complete step four, but I noticed that the training performance did not improve significantly as the epoch count increased. During training, I encountered a runtime warning about an invalid value encountered in intersection return lib.intersection(a, b, **kwargs). Eventually, when attempting to continue with step five, I received an error stating KeyError: 'ref_junction_map'. Can you please offer some assistance?

**Comment by xiannsa** [comment]
Hello, thank you for your help earlier. I was able to successfully complete step four, but I noticed that the training performance did not improve significantly as the epoch count increased. During training, I encountered a runtime warning about   invalid value encountered in intersection return lib.intersection(a, b, **kwargs).  Eventually, when attempting to continue with step five, I received an error stating KeyError: 'ref_junction_map'. Can you please offer some assistance?

**Comment by rpautrat** [comment]
Hi, did you start training using the pre-trained weights? If so, it would make sense that the training performance will not improve much on your custom dataset.

Regarding the warning, I am not sure, I would need to see the full message, but it is probably not too serious.

For step 5, did you make sure to put the option 'return_type' to 'paired_desc' in the config file of your custom dataset?

**Comment by xiannsa** [comment]
![440AR3HCUWWOCS IXN1(HTR](https://user-images.githubusercontent.com/108583027/234487963-fd11d62d-d723-4f9b-9fb9-88546af5d2c7.png)
Thank you for your message. I have already begun the fifth step of my training. Regarding the fourth step, I used my own dataset and did not use a pre-trained model from the first step. I am curious to know if the training results were very poor.

**Comment by rpautrat** [comment]
These results are actually not very poor, they are similar to what I have. The training metrics are pixel-level precision and recall, so it is very hard to get a good score in these. But the overall quality of your lines should already pretty good now. Did you try visualize them?

**Comment by xiannsa** [comment]
![_(%}V_ TR3T73DH1$QXK(08](https://user-images.githubusercontent.com/108583027/234495497-87a7429b-fee6-4cc9-9b97-f4673fcf7ff3.png)
![H5)MKQG168RTBB)SJ8I5B9E](https://user-images.githubusercontent.com/108583027/234495595-d912c79b-00a1-45b5-82a6-51fba2290754.png)
"I'm on step 5 of my training, and I used the model file from round 11 that was just trained to draw a picture. The result had more lines than when I used the wireframe.tar you provided, but there were also more mistakes. I hope that the models trained in the future will have better results."

**Comment by rpautrat** [comment]
The matching looks very good at least, but the lines could indeed be improved. Note that you can also tune the hyperparameters to make it better.

One issue here is for example the lack of junctions, which prevent from discovering too many lines. You could try reducing the parameter 'detection_thresh' of the model config file `train_full_pipeline.yaml`, so that the network can detect more junctions.

**Comment by abhiagwl4262** [comment]
@rpautrat can you just help me with .mat files for validation data ?  what is the scale the lines have been resized to ?

**Comment by rpautrat** [comment]
I am sorry, which .mat file? What do you want to validate exactly? If you follow the metrics of the paper, there should be no .mat file needed, the metrics are computed between line detections across two images warped by a homography.

**Comment by abhiagwl4262** [comment]
@rpautrat 

in this line you require a .mat file - https://github.com/cvg/SOLD2/blob/3d7bcd602ff4fb44e118d745a10251c829fa0ab1/sold2/dataset/wireframe_dataset.py#L168

How do I create one for my custom data?

**Comment by rpautrat** [comment]
Hi, you should take example on `holicity_dataset.py` and not `wireframe_dataset.py` when creating the dataloader for your own dataset. The Wireframe dataset has indeed some special features (inlcuding these .mat files coming from the original ground truth), but for a generic dataset like yours, there is no such special features. So it is better to copy the structure of `holicity_dataset.py`, where there is no such .mat file necessary.

**Comment by rpautrat** [comment]
Steps 2 and 3 go together, and you can verify their output in the notebook https://github.com/cvg/SOLD2/blob/main/notebooks/visualize_exported_dataset.ipynb

**Comment by rpautrat** [comment]
Hi, I don't understand, why do you have the exported data in this format? If you follow step 3, the output file should only have keys 'junctions' and 'line_map'. So how did you get yours?

**Comment by abhiagwl4262** [comment]
@rpautrat  If you don't recommend Wireframe then why it should not be default option of the repo. And it would be very helpful if you can explain how I can turn lcnn annotations into holicity dataset format in the readme.

**Comment by abhiagwl4262** [comment]
@rpautrat Even if I use wireframe format why I don't have 
`ref_image, ref_junctions, ref_line_map, ref_line_points` keys ?

**Comment by abhiagwl4262** [comment]
> I did try visualizing the exported dataset using above notebook -
> 
> My dataset is in wireframe format.
> 
> The keys the exported data has are - `dict_keys(['image', 'junctions', 'junction_map', 'line_map_pos', 'line_map_neg', 'heatmap_pos', 'heatmap_neg', 'valid_mask', 'file_key'])`
> 
> And throwing the error -
> 
> `KeyError: 'ref_image`

I am not able to visualize exported data of original wireframe dataset as well. I am facing same issue. Please help because You guys made it work for original wireframe dataset

**Comment by abhiagwl4262** [comment]
@rpautrat I also don't see "ref_image" key in any code related to exporting pseudo labels. Is the notebook - https://github.com/cvg/SOLD2/blob/main/notebooks/visualize_exported_dataset.ipynb correct ?

**Comment by rpautrat** [comment]
Hi @abhiagwl4262 , I think there is a confusion regarding the wireframe dataset. When we train on it, we use our own exported ground truth, we don't use the original ground truth. So yes, wireframe is the default option but with exported ground truth. The possibility to use this 'official' ground truth is a legacy and was only used during evaluation, but it is not intended for you to use it. I hope this clarifies the confusion.

The right way to use this repo is to export your own ground truth  with steps 2 and 3 (on the Wireframe dataset or any other dataset), then use this pseudo ground truth. The wireframe dataloader doesn't support the use of the 'official' ground truth and pair loading (i.e. to get this 'ref_image').

In case you are looking for it, the 'ref_image' comes from here: https://github.com/cvg/SOLD2/blob/3d7bcd602ff4fb44e118d745a10251c829fa0ab1/sold2/dataset/wireframe_dataset.py#L755
It is available when you use an exported ground truth, together with the 'paired_desc' option.

**Comment by abhiagwl4262** [comment]
@rpautrat Do I have to run step 2 and 3 with 'paired_desc' option ?

**Comment by rpautrat** [comment]
No, 'paired_desc' is only for step 5, when training the descriptor. That is why it is not activated in the default configuration.

**Comment by abhiagwl4262** [comment]
@rpautrat But you said the keys expected in the [notebook](https://github.com/cvg/SOLD2/blob/main/notebooks/visualize_exported_dataset.ipynb)  will be present when I use `paired_desc` option

**Comment by rpautrat** [comment]
Ah maybe that's where the confusion came from. The 'paired_desc' option has to be put in the field 'return_type' of the `config/wireframe_dataset.yaml` config file. This was explained for step 5, but it was probably not clear that you have to do the same for the notebook.

Steps 2 and 3 do not need this 'paired_desc' and it will anyway be ignored if you use it. So you can reuse your current results.

**Comment by abhiagwl4262** [comment]
So this notebook to visual_exported_data is of no use ?

**Comment by abhiagwl4262** [comment]
@rpautrat That would be nice. please do that. 

And Also if you can add details of how to convert lines in LCNN format(JSONs) to holicity, then repo would become real good. I can raise PR for LCNN to wireframe conversion.

**Comment by rpautrat** [comment]
I updated the notebook.

A PR with a script converting from LCNN format to holicity one would be nice, yes! Thank you in advance.

**Comment by xiannsa** [comment]
Excuse me, can I use videos to draw lines besides drawing on images in Jupyter Notebook?

**Comment by rpautrat** [comment]
If you want to make a video with lines, one option is to draw lines on each frame, then combine all frames together in a video.

**Comment by abhiagwl4262** [comment]
> I updated the notebook.
> 
> A PR with a script converting from LCNN format to holicity one would be nice, yes! Thank you in advance.

Did you check if the notebook works ?

**Comment by xiannsa** [comment]
,Excuse me ,how can i  train it   use colored original images instead of grayscale images?

**Comment by nmanhong** [comment]
@xiannsa  大哥，带我一个啊，教教我，我不会，现成的发给我看看，可以吗？

**Comment by xiannsa** [comment]
@nmanhong  你遇到什么问题了吗，训练里面的吗

**Comment by xiannsa** [comment]
@rpautrat ,Excuse me ,how can i train it use colored original images instead of grayscale images?

**Comment by rpautrat** [comment]
> > I updated the notebook.
> > A PR with a script converting from LCNN format to holicity one would be nice, yes! Thank you in advance.
> 
> @rpautrat Did you check if the notebook works ?
> 
> I got following error on `new_points = (homography @ new_points.T).T`
> 
> `ValueError: matmul: Input operand 0 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)`

Yes I ran the notebook and it works fine for me.

It seems that the homography is empty in your case, not sure why. You may want to inspect the homography generation.

**Comment by rpautrat** [comment]
> @rpautrat ,Excuse me ,how can i train it use colored original images instead of grayscale images?

In the config file, change the 'input_channel' to 3: https://github.com/cvg/SOLD2/blob/bbce15cca60ded06354106ffb66e3cb3ee23f2ad/sold2/config/train_detector.yaml#LL7C24-L7C24
In the dataloader, don't convert images to grayscale, but keep them in RGB. This should be enough to train on RGB images.

**Comment by nmanhong** [comment]
@xiannsa 大哥你是在搞训练自己的数据集吗？我也想训练自己的数据集，但是我不会搞，你搞好了 可以发给我学习学学习吗

**Comment by xiannsa** [comment]
@rpautrat    I changed the input_channel to 3 and modified the 'gray_scale: True' parameter in the holicity_dataset.yaml file to 'gray_scale: False', as you suggested. However, when I reached the fourth step of the training, I encountered the error 'ValueError: operands could not be broadcast together with remapped shapes (512, 512, 3, 1) (512, 512, 1)。Do you have any suggestions on how it ?  
![image](https://user-images.githubusercontent.com/108583027/237003469-c2a051ae-70e2-4c0b-861e-d5b16208b57f.png)
![image](https://user-images.githubusercontent.com/108583027/237003563-c2f17455-6b7b-44a7-8747-b4d654a052fa.png)

**Comment by xiannsa** [comment]
Here is the error I encountered while attempting to train with colored images using the checkpoint you provided for the second step. I hope you can give me some advice. Also, I have been able to use video for match lines, but the frame rate is a bit low. It takes 0.7 seconds per frame on my RTX 3090. I would like to know how many frames you were able to achieve in the video demonstration.
![3@5}056FJA3HD0N_XX(X H3](https://github.com/cvg/SOLD2/assets/108583027/0f1daac5-c4b7-4be0-b957-61aabbb352a4)
![image](https://github.com/cvg/SOLD2/assets/108583027/678111aa-559b-4202-a65f-2e4e3aac9fd8)

**Comment by xiannsa** [comment]
@rpautrat  Excuse me, if I lower the network depth from 4 to 3 while training, can it improve the efficiency of the model to draw lines?
![image](https://github.com/cvg/SOLD2/assets/108583027/4c4b24ca-6fdf-4598-89a3-5acfbd73bd87)

**Comment by rpautrat** [comment]
Hi, you cannot reuse the existing checkpoint for colored images, since it was trained for grayscale ones. You need to start training from scratch in your case.

Or another solution is to hack into the function loading the weights: https://github.com/cvg/SOLD2/blob/bbce15cca60ded06354106ffb66e3cb3ee23f2ad/sold2/train.py#L36-L57
and to load all the weights of the pretrained network, except the very first one, this conv1 raising the error. All the others should be identical between grayscale / rgb images.

**Comment by rpautrat** [comment]
> @rpautrat Excuse me, if I lower the network depth from 4 to 3 while training, can it improve the efficiency of the model to draw lines? ![image](https://user-images.githubusercontent.com/108583027/237263789-4c4b24ca-6fdf-4598-89a3-5acfbd73bd87.png)

This code has not been optimized for speed, it is a research code. There would be ways to make it much more efficient though. A good starting point would be to convert the line matching purely on GPU instead of partly GPU, partly CPU (e.g. here: https://github.com/cvg/SOLD2/blob/bbce15cca60ded06354106ffb66e3cb3ee23f2ad/sold2/model/line_matching.py#L280).

I would not reduce the depth of the backbone from 4 to 3, the results would probably severely decrease.

For the video, we processed the frames offline independently, then combined the frames into a video. So you can choose the frame rate as you wish with this method.

**Comment by xiannsa** [comment]
Thank you for your help. I previously tried training from scratch by modifying the 'input_channel' in 'train_detector.yaml' to 3 and training on a synthetic dataset. However, I encountered an error: 'RuntimeError: Given groups=1, weight of size..., expected input... to have 3 channels, but got 1 channel instead.' Could it be because the synthetic dataset itself is in grayscale?

**Comment by rpautrat** [comment]
Yes, of course everything is by default in grayscale, so the synthetic dataset is also in grayscale. Converting it into rgb might require more work though, especially if you want to create interesting colored texture inside the geometrical shapes...

**Comment by xiannsa** [comment]
I apologize for the interruption. May I know which parameters can be adjusted to improve the completeness of the waterline for the ship, as shown in the image? Currently, only a part of the waterline is visible.
![image](https://github.com/cvg/SOLD2/assets/108583027/77e26e3c-74a6-4092-85ce-d7129c46fde5)

**Comment by rpautrat** [comment]
Hi, here the issue is that there is no detected junction at the end of the water line, so you need to increase the number of detected junctions.

For this, you can reduce the 'detection_thresh' of the model config, and potentially also increase the 'max_num_junctions' as well, to increase the maximum number of junctions.

**Comment by xiannsa** [comment]
Thank you for your suggestion. However, I would like to know if after adjusting the parameters, do I need to retrain steps 4 and 5 and modify the corresponding parameters in the "lines_match" section to see the effect?

**Comment by rpautrat** [comment]
No, you don't need to retrain anything there. You can directly change the parameters at test time when using your already trained model.

**Comment by VitoriaCarvalho** [comment]
@rpautrat 

Hello! I'm trying to use SOLD² to detect lines in a stack of objects. In this case, I need to detect specific lines, not all straight segments in the image. Therefore, I need to train the model with my ground truth, but I am having difficulty understanding the pipeline. Is it necessary to train with synthetic data even in the case of a problem where it is not necessary to detect all straight segments present in the image?
